File: web/src/dsl/runtime/BuiltinRegistry.ts
import type { DslValue } from './DslValue'
import type { Arguments } from './Arguments'
import type { Environment } from './Environment'
import { getDateFunctions } from '../builtins/DateFunctions'
import { getCharacterConstants } from '../builtins/CharacterConstants'
import { getArithmeticFunctions } from '../builtins/ArithmeticFunctions'
import { getComparisonFunctions } from '../builtins/ComparisonFunctions'
import { getPatternFunctions } from '../builtins/PatternFunctions'
import { getNoteFunctions } from '../builtins/NoteFunctions'
import { getListFunctions } from '../builtins/ListFunctions'
import { getSortConstants } from '../builtins/SortConstants'
import { getActionFunctions } from '../builtins/ActionFunctions'

export interface BuiltinFunction {
  name: string
  isDynamic?: boolean
  call: (args: Arguments, env: Environment) => DslValue | Promise<DslValue>
}

const functions = new Map<string, BuiltinFunction>()
let initialized = false

function ensureInitialized(): void {
  if (initialized) return
  initialized = true

  const allFunctions: BuiltinFunction[] = [
    ...getDateFunctions(),
    ...getCharacterConstants(),
    ...getArithmeticFunctions(),
    ...getComparisonFunctions(),
    ...getPatternFunctions(),
    ...getNoteFunctions(),
    ...getListFunctions(),
    ...getSortConstants(),
    ...getActionFunctions(),
  ]

  for (const fn of allFunctions) {
    functions.set(fn.name, fn)
  }
}

export const BuiltinRegistry = {
  get(name: string): BuiltinFunction | null {
    ensureInitialized()
    return functions.get(name) ?? null
  },

  has(name: string): boolean {
    ensureInitialized()
    return functions.has(name)
  },

  isDynamic(name: string): boolean {
    ensureInitialized()
    return functions.get(name)?.isDynamic ?? false
  },

  allNames(): Set<string> {
    ensureInitialized()
    return new Set(functions.keys())
  },
}

File: app/src/main/java/org/alkaline/taskbrain/dsl/runtime/BuiltinRegistry.kt
package org.alkaline.taskbrain.dsl.runtime

import org.alkaline.taskbrain.dsl.builtins.ActionFunctions
import org.alkaline.taskbrain.dsl.runtime.values.DslValue
import org.alkaline.taskbrain.dsl.builtins.ArithmeticFunctions
import org.alkaline.taskbrain.dsl.builtins.CharacterConstants
import org.alkaline.taskbrain.dsl.builtins.ComparisonFunctions
import org.alkaline.taskbrain.dsl.builtins.DateFunctions
import org.alkaline.taskbrain.dsl.builtins.ListFunctions
import org.alkaline.taskbrain.dsl.builtins.NoteFunctions
import org.alkaline.taskbrain.dsl.builtins.PatternFunctions
import org.alkaline.taskbrain.dsl.builtins.SortConstants

/**
 * A builtin function that can be called from Mindl.
 *
 * @property name The function name used to call it
 * @property isDynamic True if this function can return different results on each call
 *                     (e.g., `now`, `date`, `time`). Static functions always return the
 *                     same result for the same inputs (e.g., `iso8601`, `qt`).
 * @property call The implementation that takes arguments and an environment
 */
data class BuiltinFunction(
    val name: String,
    val isDynamic: Boolean = false,
    val call: (args: Arguments, env: Environment) -> DslValue
)

/**
 * Registry of all builtin functions available in Mindl.
 *
 * Adds arithmetic functions.
 * Adds note functions (find).
 * Adds list functions (sort, first).
 */
object BuiltinRegistry {
    private val functions = mutableMapOf<String, BuiltinFunction>()

    init {
        // Register all builtin modules
        DateFunctions.register(this)
        CharacterConstants.register(this)
        ArithmeticFunctions.register(this)
        ComparisonFunctions.register(this)
        PatternFunctions.register(this)
        NoteFunctions.register(this)
        ListFunctions.register(this)
        SortConstants.register(this)
        ActionFunctions.register(this)
    }

    /**
     * Register a builtin function.
     */
    fun register(function: BuiltinFunction) {
        functions[function.name] = function
    }

    /**
     * Look up a function by name.
     * @return The function, or null if not found
     */
    fun get(name: String): BuiltinFunction? = functions[name]

    /**
     * Check if a function exists.
     */
    fun has(name: String): Boolean = functions.containsKey(name)

    /**
     * Check if a function is dynamic (can return different results each call).
     */
    fun isDynamic(name: String): Boolean = functions[name]?.isDynamic ?: false

    /**
     * Get all registered function names.
     */
    fun allNames(): Set<String> = functions.keys.toSet()
}
